src/as_execution/mod.rs

```diff
@@ -60,36 +59,32 @@ impl ASModule {
 }
 
 pub(crate) fn init_engine(limit: u64, gas_costs: GasCosts) -> Engine {
-    // We use the Singlepass compiler because it is fast and adapted to blockchains
-    // See https://docs.rs/wasmer-compiler-singlepass/latest/wasmer_compiler_singlepass/
-    let mut compiler_config = Singlepass::new();
+    // Use Cranelift (in opposition to Singlepass)
+    // Because of module caching we can run longer compilations to have better optimizations
+    // Executions are happening way more often than compilations hence that choice
+    let mut compiler_config = Cranelift::new();
 
-    // Turning-off sources of potential non-determinism,
-    // see https://github.com/WebAssembly/design/blob/037c6fe94151eb13e30d174f5f7ce851be0a573e/Nondeterminism.md
-
-    // Turning-off in the compiler:
-
-    // Canonicalize NaN.
+    // Canonicalize NaN & turn off sources of potential non-determinism:
+    // * threads
+    // * SIMD
+    // * experimental features
+    // References:
+    // * https://github.com/WebAssembly/design/blob/390bab47efdb76b600371bcef1ec0ea374aa8c43/Nondeterminism.md
+    // * https://github.com/WebAssembly/proposals
     compiler_config.canonicalize_nans(true);
-
-    // Default: Turning-off all wasmer feature flags
-    // Exception(s):
-    // * bulk_memory:
-    //   * https://docs.rs/wasmer/latest/wasmer/struct.Features.html: now fully standardized - wasm 2.0
-    //   * See also: https://github.com/paritytech/substrate/issues/12216
     const FEATURES: Features = Features {
```

Comment:
This enables some new WASM feature like multi_value & reference_types but why? Are we sure this does not impact determinism?

Reply:
Multi value and reference types were disabled because Singlepass compiler is not able to handle them, not because they could hurt the determinism. I've read the documentation on this and they are not referenced as non-deterministic  nor experimental features, we're good on that.

Reply:
I would be really conservative on that. Maybe I can double check if you really want to enable them + check with @damip . From a quick look, this is disabled in Near code...

Reply:
See https://github.com/massalabs/massa-sc-runtime/pull/232#discussion_r1105646160

Reply:
We want the same behavior (not necessarily the same perf) between singlepass and cranelift-compiled binaries. So if one compiler supports a feature another compiler doesn't, then that feature should be disabled to avoid behavioral discrepancy